crates/ironrdp-client/src/network_cllient.rs

```diff
@@ -0,0 +1,119 @@
+use std::net::{IpAddr, Ipv4Addr};
+use std::{future::Future, pin::Pin};
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use tokio::net::{TcpStream, UdpSocket};
+
+use ironrdp::connector::{custom_err, general_err, ConnectorResult};
+use reqwest::Client;
+use url::Url;
+
+use ironrdp_tokio::AsyncNetworkClient;
+pub(crate) struct AsyncTokioNetworkClient {
+    client: Option<Client>,
+}
+impl AsyncNetworkClient for AsyncTokioNetworkClient {
+    fn send<'a>(
+        &'a mut self,
+        request: &'a sspi::generator::NetworkRequest,
+    ) -> Pin<Box<dyn Future<Output = ConnectorResult<Vec<u8>>> + 'a>> {
+        Box::pin(async move {
+            match &request.protocol {
+                sspi::network_client::NetworkProtocol::Tcp => self.send_tcp(&request.url, &request.data).await,
+                sspi::network_client::NetworkProtocol::Udp => self.send_udp(&request.url, &request.data).await,
+                sspi::network_client::NetworkProtocol::Http | sspi::network_client::NetworkProtocol::Https => {
+                    self.send_http(&request.url, &request.data).await
+                }
+            }
+        })
+    }
+}
+
+impl AsyncTokioNetworkClient {
+    pub(crate) fn new() -> Self {
+        Self { client: None }
+    }
+}
+
+impl AsyncTokioNetworkClient {
+    async fn send_tcp(&self, url: &Url, data: &[u8]) -> ConnectorResult<Vec<u8>> {
+        let addr = format!("{}:{}", url.host_str().unwrap_or_default(), url.port().unwrap_or(88));
+        let mut stream = TcpStream::connect(addr)
+            .await
+            .map_err(|e| Error::new(ErrorKind::NoAuthenticatingAuthority, format!("{:?}", e)))
+            .map_err(|e| custom_err!("sending KDC request over TCP ", e))?;
+
+        stream
+            .write(data)
+            .await
+            .map_err(|e| Error::new(ErrorKind::NoAuthenticatingAuthority, format!("{:?}", e)))
+            .map_err(|e| custom_err!("Sending KDC request over TCP ", e))?;
+
+        let len = stream
+            .read_u32()
+            .await
+            .map_err(|e| Error::new(ErrorKind::NoAuthenticatingAuthority, format!("{:?}", e)))
+            .map_err(|e| custom_err!("Sending KDC request over TCP ", e))?;
+
+        let mut buf = vec![0; len as usize + 4];
+        buf[0..4].copy_from_slice(&(len.to_be_bytes()));
+
+        stream
+            .read_exact(&mut buf[4..])
+            .await
+            .map_err(|e| Error::new(ErrorKind::NoAuthenticatingAuthority, format!("{:?}", e)))
+            .map_err(|e| custom_err!("Sending KDC request over TCP ", e))?;
+
+        Ok(buf)
+    }
+
+    async fn send_udp(&self, url: &Url, data: &[u8]) -> ConnectorResult<Vec<u8>> {
+        let port =
+            portpicker::pick_unused_port().ok_or_else(|| general_err!("No ports free for sending udp request"))?;
+
+        let udp_socket = UdpSocket::bind((IpAddr::V4(Ipv4Addr::LOCALHOST), port))
+            .await
+            .map_err(|e| custom_err!("Cannot bind udp socket", e))?;
```

Comment:
Remove the dependency on `portpicker`, and ask for a port assigned by the operating system:
```suggestion
        let udp_socket = UdpSocket::bind((IpAddr::V4(Ipv4Addr::LOCALHOST), 0))
            .await
            .map_err(|e| custom_err!("Cannot bind udp socket", e))?;
```